vendor/mcmc/mcmc.c

```diff
@@ -358,32 +459,272 @@ static int _mcmc_parse_response(mcmc_ctx_t *ctx, mcmc_resp_t *r) {
+                if (sum > lim) {
+                    return MCMC_TOKTO_ERANGE;
+                }
+                sum = sum * 10 + num;
+            }
+        } else {
+            return MCMC_TOKTO_EINVALID;
+        }
+        pos++;
+    }
+    *out = sum;
+    return MCMC_OK;
 }
 
-// Allow returning this dynamically based on options set.
-// FIXME: it might be more flexible to call this after mcmc_connect()...
-// but this is probably more convenient for the caller if it's less dynamic.
-size_t mcmc_min_buffer_size(int options) {
-    return MIN_BUFFER_SIZE;
+MCMC_STATIC int mcmc_tokto64(const char *t, size_t len, int64_t *out) {
+    int64_t sum = 0;
+    const char *pos = t;
+    int is_sig = 0;
+    if (len > MCMC_TOKTO64_MAX) {
+        return MCMC_TOKTO_ELONG;
+    }
+    // If we're negative the first character must be -
+    if (pos[0] == '-') {
+        len--;
+        pos++;
+        is_sig = 1;
+    }
+    while (len--) {
+        char num = pos[0] - '0';
+        if (num > -1 && num < 10) {
+            if (is_sig) {
+                int64_t lim = (INT64_MIN + num) / 10;
+                if (sum < lim) {
+                    return MCMC_TOKTO_ERANGE;
+                }
+                sum = sum * 10 - num;
+            } else {
+                int64_t lim = (INT64_MAX - num) / 10;
+                if (sum > lim) {
+                    return MCMC_TOKTO_ERANGE;
+                }
+                sum = sum * 10 + num;
+            }
+        } else {
+            return MCMC_TOKTO_EINVALID;
+        }
+        pos++;
+    }
+    *out = sum;
+    return MCMC_OK;
+}
+
+// END INTERNAL FUNCTIONS
+
+// Context-less bare API.
+
+//MCMC_STATIC int _mcmc_tokenize_meta(mcmc_tokenizer_t *t, const char *line, size_t len, const int mstart, const int max) {
+
+// TODO: ideally this does a macro test of t->ntokens before calling a
+// function. are compilers smart enough that I don't have to do this tho?
+int mcmc_tokenize_res(const char *l, size_t len, mcmc_tokenizer_t *t) {
+    if (t->ntokens == 0) {
+        return _mcmc_tokenize_meta(t, l, len, 1, MCMC_PARSER_MAX_TOKENS-1);
+    }
+    return 0;
+}
+
+const char *mcmc_token_get(const char *l, mcmc_tokenizer_t *t, int idx, int *len) {
+    if (idx > 0 && idx < t->ntokens) {
+        return _mcmc_token(l, t, idx, len);
+    } else {
+        return NULL;
+    }
+}
+
+#define X(n, p, c) \
+    int n(const char *l, mcmc_tokenizer_t *t, int idx, p *val) { \
+        if (idx > 0 && idx < t->ntokens) { \
+            int tlen = 0; \
+            const char *tok = _mcmc_token(l, t, idx, &tlen); \
+            return c(tok, tlen, val); \
+        } else { \
+            return MCMC_ERR; \
+        } \
+    }
+
+X(mcmc_token_get_u32, uint32_t, mcmc_toktou32)
+X(mcmc_token_get_u64, uint64_t, mcmc_toktou64)
+X(mcmc_token_get_32, int32_t, mcmc_tokto32)
+X(mcmc_token_get_64, int64_t, mcmc_tokto64)
+
+#undef X
+
+int mcmc_token_has_flag(const char *l, mcmc_tokenizer_t *t, char flag) {
```

Comment:
I've forgotten the weird rules around signed vs unsigned chars and what platforms have what. might be worth doing this differently, but probably not now.